File: Week8/clock.py
import sys
from heapq import heappush
"""
  Brian Grenier
  1545276
  bgrenier

  List any resources you used below (eg. urls, name of the algorithm from our code archive).
  Remember, you are permitted to get help with general concepts about algorithms
  and problem solving, but you are not permitted to hunt down solutions to
  these particular problems!


  List any classmate you discussed the problem with. Remember, you can only
  have high-level verbal discussions. No code should be shared, developed,
  or even looked at in these chats. No formulas or pseudocode should be
  written or shared in these chats.

  N/A

  By submitting this code, you are agreeing that you have solved in accordance
  with 
"""
def diff(a, b):
    return abs(((a-b) + 180000) % (360000 - 180000))

def pi(p, nex):
    nex[0] = -1
    k = -1
    j = 0
    while(j < len(p)-1):
        if k == -1 or p[j] == p[k]:
            k+=1
            j+=1
            if (p[j] != p[k]):
                nex[j] = k
            else:
                nex[j] = k
        else:
            k = nex[k]

def kmp(s, p):
    i = 0
    j = 0
    nex = [0] * 200005
    
    pi(p, nex)

    while (i < len(s) and j < len(p)):
        if j == -1 and s[i] == p[j]:
            i += 1
            j += 1
        else:
            j = nex[j]

    if (j == len(p)):
        return True
    return False



if __name__ == '__main__':
    n = int(input())

    a = list(map(int, input().split()))
    b = list(map(int, input().split()))

    a.sort()
    b.sort()

    c = [0] *n
    d = [0] *n

    for i in range(1, n):
        c[i-1] = diff(a[i], a[i-1])
        d[i-1] = diff(b[i], b[i-1])
    c[n-1] = diff(a[0], a[n-1])
    d[n-1] = diff(b[0], b[n-1])

    c += c


    if kmp(c, d):
        print('possible')
    else:
        print('impossible')

File: Week8/clock.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;

/*
  Brian Grenier
  1545276
  bgrenier

  List any resources you used below (eg. urls, name of the algorithm from our code archive).
  Remember, you are permitted to get help with general concepts about algorithms
  and problem solving, but you are not permitted to hunt down solutions to
  these particular problems!

  https://github.com/kth-competitive-programming/kactl/blob/master/content/strings/KMP.h

  List any classmate you discussed the problem with. Remember, you can only
  have high-level verbal discussions. No code should be shared, developed,
  or even looked at in these chats. No formulas or pseudocode should be
  written or shared in these chats.

  N/A

  By submitting this code, you are agreeing that you have solved in accordance
  with the collaboration policy in CMPUT 403.
*/

int mod(int a, int b) {
    return a - floor(a/b) * b;
}

int diff(int a1, int a2) {
    int temp = a1 - a2;
    return abs(mod((temp + 180000), 360000 - 180000));
}

vi pi(const string& s) {
	vi p(sz(s));
	rep(i,1,sz(s)) {
		int g = p[i-1];
		while (g && s[i] != s[g]) g = p[g-1];
		p[i] = g + (s[i] == s[g]);
	}
	return p;
}

bool match(const string& s, const string& pat) {
	vi p = pi(pat + '\0' + s), res;
	rep(i,sz(p)-sz(s),sz(p))
		if (p[i] == sz(pat)) res.push_back(i - 2 * sz(pat));
	return res.size() == s.size();
}

int main() {
    int n;

    cin >>n;

    int c1[n];
    int c2[n];
    for (int i = 0; i < n; ++i) {
        cin >> c1[i];
    }

    for (int i = 0; i < n; ++i) {
        cin >> c2[i];
    }

    sort(c1, c1+n);
    sort(c2, c2+n);

    vi a(n*2);
    vi b(n);

    for (int i = 1; i < n; ++i) {
        a[i-1] = c1[i] - c1[i-1];
        b[i-1] = c2[i] - c2[i-1];
    }
    a[n-1] = 360000 + c1[0] - c1[n-1];
    b[n-1] = 360000 + c2[0] - c2[n-1];
    for(int i=0;i<n;i++) a[i+n]=a[i];
    if (match(string(a.begin(), a.end()), string(b.begin(), b.end()))) {
        cout << "possible" << endl;
    } else {
        cout << "impossible" << endl;
    }
}